R/preprocessing.R

```diff
@@ -646,7 +646,70 @@ Load10X_Spatial <- function (
   return(object)
 }
 
+#' Load a 10x Genomics Single Cell Experiment into a \code{Seurat} object
+#'
+#' @inheritParams Read10X
+#' @inheritParams SeuratObject::CreateSeuratObject
+#' @param data.dir Directory containing the H5 file specified by \code{filename}
+#' and the image data in a subdirectory called \code{spatial}
+#' @param filename Name of H5 file containing the feature barcode matrix
+#' @param filter.matrix Only keep spots that have been determined to be over
+#' tissue
+#' @param to.upper Converts all feature names to upper case. This can provide an
+#' approximate conversion of mouse to human gene names which can be useful in an
+#' explorative analysis. For cross-species comparisons, orthologous genes should
+#' be identified across species and used instead.
+#' @param ... Arguments passed to \code{\link{Read10X_h5}}
+#'
+#' @return A \code{Seurat} object
+#'
+#'
+#' @export
+#' @concept preprocessing
+#'
+#' @examples
+#' \dontrun{
+#' data_dir <- 'path/to/data/directory'
+#' list.files(data_dir) # Should show filtered_feature_bc_matrix.h5
+#' Load10X(data.dir = data_dir)
+#' }
+#'
+Load10x <- function(data.dir, filename = "filtered_feature_bc_matrix.h5",
+                    assay = "RNA", filter.matrix = TRUE, to.upper = FALSE, ...) {
+
+  if (length(data.dir) > 1) {
+    stop("`data.dir` expects a single directory path but received multiple values.")
+  }
+  if (!file.exists(data.dir)) {
+    stop(paste0("No such file or directory: '", data.dir, "'"))
+  }
+
+
+  filename <- list.files(data.dir, filename, full.names = FALSE, recursive = FALSE)
+  counts.path <- file.path(data.dir, filename)
+  if (!file.exists(counts.path)) {
+    stop(paste0("File not found: '", counts.path, "'"))
+  }
 
+  counts <- Read10X_h5(counts.path, ...)
+  if (is.list(counts)) {
+    object <- Reduce(function(x, y) merge(x, y, add.cell.ids = names(counts)),
+                     lapply(names(counts), function(name) CreateSeuratObject(counts[[name]], assay = name)))
+  } else {
+    object <- CreateSeuratObject(counts, assay = assay)
+  }
+
+  cell_types_path <- file.path(data.dir, "cell_types", "cell_types.csv")
+  if (file.exists(cell_types_path)) {
+    cell.types <- read.csv(cell_types_path)
+    object@meta.data <- dplyr::left_join(tibble::rownames_to_column(object@meta.data, "barcode"),
```

Comment:
I often do exactly this in my analysis scripts (barcodes to column, left join metadata, barcodes to rownames), but using the dplyr left join here introduces another dependency. Not a problem for me, but some other people may want to avoid this and use `base::merge`

Reply:
... or `AddMetaData`